src/ripple/protocol/tests/Seed_test.cpp

```diff
@@ -0,0 +1,215 @@
+        {
+            uint128 src;
+            beast::rngfill (src.data(), src.size(), g);
+            Seed const seed (src);
+            expect (memcmp (seed.data(), src.data(), src.size()) == 0);
+        }
+    }
+
+    std::string testPassphrase(std::string passphrase)
+    {
+        auto const seed1 = generateSeed (passphrase);
+        auto const seed2 = parseBase58<Seed>(toBase58(seed1));
+
+        expect (static_cast<bool>(seed2));
+        expect (seed1 == seed2);
+        return toBase58(seed1);
+    }
+
+    void testPassphrase()
+    {
+        testcase ("generation from passphrase");
+        expect (testPassphrase ("masterpassphrase") ==
+            "snoPBrXtMeMyMHUVTgbuqAfg1SUTb");
+        expect (testPassphrase ("Non-Random Passphrase") ==
+            "snMKnVku798EnBwUfxeSD8953sLYA");
+        expect (testPassphrase ("cookies excitement hand public") ==
+            "sspUXGrmjQhq6mgc24jiRuevZiwKT");
+    }
+
+    void testBase58()
+    {
+        testcase ("base58 operations");
+
+        // Success:
+        expect (parseBase58<Seed>("snoPBrXtMeMyMHUVTgbuqAfg1SUTb"));
+        expect (parseBase58<Seed>("snMKnVku798EnBwUfxeSD8953sLYA"));
+        expect (parseBase58<Seed>("sspUXGrmjQhq6mgc24jiRuevZiwKT"));
+
+        // Failure:
+        expect (!parseBase58<Seed>(""));
+        expect (!parseBase58<Seed>("sspUXGrmjQhq6mgc24jiRuevZiwK"));
+        expect (!parseBase58<Seed>("sspUXGrmjQhq6mgc24jiRuevZiwKTT"));
+        expect (!parseBase58<Seed>("sspOXGrmjQhq6mgc24jiRuevZiwKT"));
+        expect (!parseBase58<Seed>("ssp/XGrmjQhq6mgc24jiRuevZiwKT"));
+    }
+
+    void testRandom()
+    {
+        testcase ("random generation");
+
+        for (int i = 0; i < 32; i++)
+        {
+            auto const seed1 = randomSeed ();
+            auto const seed2 = parseBase58<Seed>(toBase58(seed1));
+
+            expect (static_cast<bool>(seed2));
+            expect (seed1 == seed2);
+        }
+    }
+
+    void testKeypairGenerationAndSigning ()
+    {
+        std::string const message = "http://www.ripple.com";
+
+        {
+            testcase ("Node keypair generation & signing (secp256k1)");
+
+            auto const secretKey = generateSecretKey (
+                KeyType::secp256k1, generateSeed ("masterpassphrase"));
+            auto const publicKey = derivePublicKey (
+                KeyType::secp256k1, secretKey);
+
+            expect (toBase58(TokenType::TOKEN_NODE_PUBLIC, publicKey) ==
+                "n94a1u4jAz288pZLtw6yFWVbi89YamiC6JBXPVUj5zmExe5fTVg9");
+            expect (toBase58(TokenType::TOKEN_NODE_PRIVATE, secretKey) ==
+                "pnen77YEeUd4fFKG7iycBWcwKpTaeFRkW2WFostaATy1DSupwXe");
+            expect (to_string(calcNodeID(publicKey)) ==
+                "7E59C17D50F5959C7B158FEC95C8F815BF653DC8");
+
+            auto sig = sign (publicKey, secretKey, makeSlice(message));
+            expect (sig.size() != 0);
+            expect (verify (publicKey, makeSlice(message), sig));
```

Comment:
This could benefit from negative test cases as well. (Corrupt `sig`, and `expect(!verify...`.)